packages/unified-selection/src/unified-selection/HiliteSetProvider.ts

```diff
@@ -105,43 +125,72 @@ class HiliteSetProviderImpl implements HiliteSetProvider {
       return cachedType;
     }
 
-    const keyClass = await this.getClass(key.className);
-    if (!keyClass) {
-      this._classRelationCache.set(key.className, "unknown");
-      return "unknown";
-    }
+    const promise = this.getTypeImpl(key).then((res) => {
+      // Update the cache with the result of the promise.
+      this._classRelationCache.set(key.className, res);
+      return res;
+    });
+
+    // Add the promise to cache to prevent `getTypeImpl` being called multiple times.
+    this._classRelationCache.set(key.className, promise);
+    return promise;
+  }
 
+  private async getTypeImpl(key: SelectableInstanceKey): Promise<InstanceIdType> {
+    const keyClass = await this.getClass(key.className);
```

Comment:
Do you think using `BaseClassChecker` that we have in `hierarchies` package would be helpful here?

Reply:
It probably would be useful, but should `unified-selection` now have `hierarchies` as a dependency? Or do you suggest we create another package?

Reply:
`hierarchies` should definitely not be a dependency for `unified-selection`. We'll need to come up with a way to share such code between packages, but in the meantime I suggest we simply have a hard copy of the thing.

Reply:
`BaseClassChecker` has a little bit different purpose. It caches, whether one class is base of another. Here we don't care about this information, we only want to know class names by instance ID. I think, we should leave it as it is now, or I can extract this logic to a separate class if you think it might be useful

Reply:
Oh yes, my bad, you are right. But still, current implementation caches `instanceIdType` which helps us prevent doing this for each selectable:
```ts
    const checkType = async (idType) => await baseClassChecker.is(selectable, idType) ? idType : undefined);
    return (
      (await checkType("subject")) ??
      (await checkType("model")) ??
      (await checkType("category")) ??
      (await checkType("subCategory")) ??
      (await checkType("functionalElement", "Functional")) ??
      (await checkType("groupInformationElement")) ??
      (await checkType("geometricElement")) ??
      (await checkType("element")) ??
      "unknown"
    );
 ```
 Base class checker doesn't help in this regard

Reply:
okay, I now see that we already have `_classRelationCache` for storing `class name -> end result` map, so I agree there's no need to another, intermediate, cache